sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/V1Beta3.java

```diff
@@ -247,9 +253,44 @@ public String getNamespace() {
       return namespace;
     }
 
+    /**
+     * The composite {@link PTransform} for reading from datastore.
+     *
+     * It involves the following steps,
+     * 1. Create a Singleton of the user provided {@code query} and apply a {@link ParDo}
+     *    that splits the query into {@code numQuerySplits}. If {@code numQuerySplits} is 0 then
+     *    datastore is queried to get the estimated size, and use a default bundle size
+     *    to determine the number of splits.
+     * 2. The resulting {@link PCollection<Query>} is randomly distributed into buckets through a
+     *    {@link GroupByKey} operation.
+     * 3. Apply a {@link ParDo} that reads {@link Entity} from datastore for each of the above
+     *    generated queries.
+     *
+     * Note: This allows for dynamic rebalancing to happen after the {@link GroupByKey} operation,
+     * and avoids the complexity of implementing a {@link org.apache.beam.sdk.io.BoundedSource}
+     * to support dynamic splitting.
+     */
     @Override
     public PCollection<Entity> apply(PBegin input) {
-      return input.apply(org.apache.beam.sdk.io.Read.from(getSource()));
+      V1Beta3Options v1Beta3Options = V1Beta3Options.from(getProjectId(), getQuery(),
+          getNamespace());
+
+      PCollection<Query> queries = input
+          .apply(Create.of(query))
```

Comment:
`Create.of(query).withCoder(ProtoCoder.of(Query.class))` -- coder goes in the `Create`.